mne/preprocessing/GradientRemover.py

```diff
@@ -0,0 +1,258 @@
+
+        Returns
+        -------
+        np.ndarray of shape (channels, tr_timepoints)
+            Represents the template data at the given TR.
+
+        Raises
+        ------
+        ValueError if an invalid TR index is supplied.
+        """
+        detrended = self.get_tr_detrended(n)
+        template = self.get_tr_template(n)
+        return detrended - template
+
+    def correct(self):
+        """Generate the gradient-corrected data.
+
+        Returns
+        -------
+        None
+            Returns None but fills in gradient-corrected data to this
+            object.
+        """
+        corrected = self._data.copy()
+        for tr in range(self.n_tr):
+            this_start, this_end = self._tr_bounds(tr)
+            corrected[:, this_start:this_end] = self.get_tr_corrected(tr)
+        self._corrected = corrected
+        return corrected
+
+    def _valid_window(window):
+        if isinstance(window, int):
+            if not window % 2 == 0:
+                raise ValueError(
+                    f"Integer windows must be even (received {window})."
+                )
+            window = (window // 2, window // 2)
+        elif isinstance(window, tuple):
+            if not len(window) == 2:
+                raise ValueError(
+                    "Tuple windows must contain 2 elements "
+                    f"(received {window})."
+                )
+        else:
+            raise TypeError(
+                "Window must be a positive, even integer or a tuple of "
+                "size 2 containing a positive integer."
+                "(Received {window})."
+            )
+        if window[0] < 0 or window[1] < 0:
+            raise ValueError(
+                "Window must contain a positive integer. "
+                f"(Received {window})."
+            )
+        if window[0] == 0 and window[1] == 0:
+            raise ValueError(
+                "Window must contain a positive integer. "
+                f"(Received {window})."
+            )
+        return window
+
+    def _valid_tr_events(tr_events):
+        # Check to make sure TRs are evenly spaced
+        if len(tr_events.shape) == 2:
+            if tr_events.shape[1] == 3:
+                tr_events = tr_events[:, 0]
+            else:
+                raise ValueError(
+                    "TRs must be a 1D array or a (N, 3) ndarray from mne. "
+                    f"Received array of shape {tr_events.shape}."
+                )
+        elif len(tr_events.shape) != 1:
+            raise ValueError(
+                "TRs must be a 1D array or a (N, 3) ndarray from mne. "
+                f"Received array of shape {tr_events.shape}."
+            )
+        unique = np.unique(np.diff(tr_events))
+        if len(unique) != 1:
+            raise ValueError(
+                "TR spacings are not consistent; the following unique "
+                f"distances were present: {unique}."
+            )
```

Comment:
This is problematic for my test data and I guess many other datasets too. MRI gradients will not always occur on an exact sample. Hence, the event will be rounded to the nearest sample, sometimes rounded down, sometimes rounded up. For example in my data, the interval between gradient onsets is sometimes 306 and sometimes 307 samples.

Reply:
I believe that @jbteves has now addressed this by getting the  `mode( )` TR different and then doing the necessary shifting of samples.